src/python/pants/backend/jvm/targets/scala_library.py

```diff
@@ -82,3 +92,59 @@ def java_sources(self):
         raise TargetDefinitionException(self, 'No such java target: {}'.format(spec))
       targets.append(target)
     return targets
+
+  @property
+  def scalac_plugins(self):
+    """The names of compiler plugins to use when compiling this target with scalac.
+    :return: See constructor.
+    :rtype: list of strings.
+    """
+    if self._scoverage_instance.get_options().enable_scoverage:
+      # Prevent instrumenting generated targets and targets in blacklist.
+      if self.identifier.startswith(".pants.d.gen") or self._scoverage_instance.is_blacklisted(self):
+        return self.payload.scalac_plugins
+
+      scalac_plugins = self.payload.scalac_plugins
+      if scalac_plugins:
+        scalac_plugins.append(SCOVERAGE)
+      else:
+        scalac_plugins = [SCOVERAGE]
+      return scalac_plugins
+
+    return self.payload.scalac_plugins
+
+  @property
+  def scalac_plugin_args(self):
+    """Map from scalac plugin name to list of args for that plugin.
+    :return: See constructor.
+    :rtype: map from string to list of strings.
+    """
+    if self._scoverage_instance.get_options().enable_scoverage:
+      scalac_plugin_args = self.payload.scalac_plugin_args
```

Comment:
These options should be added to the plugin args in the constructor, rather than here. That would mean that they end up in the payload, and end up in the fingerprint for the target.